Introduction of convex polygons in geometry + benchmarks

BlackMinou opened the pull request:
Nit implementation of the convex polygon data structure.

It's based on the java implementation done during the data structures course. 

There also are benchmarks for the Nit implementation and the Java implementation to compare them

R4PaSs commented:
Am I the only one who cant comment on the code ?

privat commented:
I will also need the signoff of @Johann673 on commits with his code

R4PaSs commented:
@privat wrong Johan, that's @Djomanix you meant here

R4PaSs commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
```
Why the double blank line ?

R4PaSs commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+
+import points_and_lines
+
+
```
Same here

R4PaSs commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+	# Vertices of this polygon
+	var points = new Array[Point[Float]]
+
+	# Init this polygon with a list of at least 3 vertices
```
I feel that the comment is not clear enough, maybe add a REQUIRE instead of the 3 vertices part

R4PaSs commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+	end
+
+	# Sort the vertices in counter clockwise order
+	# ~~~
```
I don't know if this is requirement, but as a personal preference, I like having a blank line between the nitunits and the synopsis

R4PaSs commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+	end
+end
+
+private class Projection
```
I'm not sure I understand the use of this class, though private, a small synopsis to explain its use might be nice for future maintainers

R4PaSs commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+end
+
+# AntiClockSort
+private class AntiClockSort
```
Since it super `PolygonSorter`, why not name it `AntiClockSorter` ?

R4PaSs commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+	end
+end
+
+# ClockSort
```
I'm not a specialist, but, I'm pretty sure this is not a good synopsis

R4PaSs commented:
Aside from what I said (and the Java/Bash part which is TL;DR), :shipit:

xymus commented on benchmarks/polygons/nit/bench_polygon.nit:
```diff
@@ -0,0 +1,141 @@
+#This file is part of NIT (http://www.nitlanguage.org).
+#
+# Copyright 2014 Romain Chanoir <romain.chanoir@viacesi.fr>
```
2014? and you can add the missing space to the first line while you're at it.

xymus commented on benchmarks/polygons/nit/bench_polygon.nit:
```diff
@@ -0,0 +1,141 @@
+	#	var point = new Point[Float](theta.cos * radius, theta.sin * radius)
+	#	points.add(point)
+	# end
+	# return points
```
Commented code, and some parameters are unused `p` and `radius`. You will also get n+1 points using `[0..n]`, you might prefer `n.times`.

xymus replied:
Opps, only `p` is unused, `radius` is alright.

xymus commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+		var x = new Array[Float]
+		for p in points do
+			x.add(p.x)
+		end
```
You can use an array comprehension for this kind of operation:
~~~
return [for p in points do p.x]
~~~

xymus commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+	end
+
+	# Get axes for the intersection test
+	private fun get_axes: Array[Point[Float]] do
```
What is an axe in this context? From the code it looks like the vector between each points, is that right?

xymus replied:
Also, there's no need for the `get_` part in the Nit version.

xymus commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+		var axes1 = get_axes
+		var axes2 = other.get_axes
+		for i in [0..axes1.length[ do
+			var axis = axes1[i]
```
You should use a standard loop:
~~~
for axis in axes1 do
~~~

xymus commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+		return projection
+	end
+
+	# Check if this polygon is convex
```
When returning a boolean we try to use a question, it is often much clearer: `Is this polygon convex?`

xymus commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+			next = points[i]
+			if turn_left(prev, curr, next) != is_ccw then return false
+		end
+		print points.length
```
dd

xymus commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+	# var arr = new Array[Point[Float]].with_items(p1, p2, p3, p4)
+	# var poly = new ConvexPolygon.with_vertices(arr)
+	# poly.anti_clock_sort
+	# assert poly.is_ccw
```
Is `is_ccw` is the validation to `anti_clock_sort`? Someone could write `if not in_ccs then anti_clock_sort`? If so, you should use more similar names, maybe use `sort_ccw`.

BlackMinou replied:
`is_ccw` is just an indication to know the order of the vertices in the polygon, it's used in the other algorithms. It supposes that the polygon's vertices are already sorted.

xymus commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+		end
+	end
+
+	# Add a vertice to the polygon
```
*vertex

xymus commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+end
+
+
+# Return the convex hull of the set of points represented by `points`
```
You can do a better use of the parameter name: Get the convex hull of the set of `points`

xymus commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,428 @@
+# var p3 = new Point[Float](0.0, 5.0)
+# assert turn_left(p1, p2, p3)
+# ~~~
+fun turn_left(p1, p2, p3: Point[Float]): Bool do
```
This could probably be private.

BlackMinou commented:
Reroll : 
* Added @Djomanix on the commit with his code
* changed the name of the sorters and the methods used to sort the polygon
* fixed the year in the license
* improved some documentation

xymus commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,426 @@
+	# Vertices of this polygon
+	var points = new Array[Point[Float]]
+
+	# Init this polygon with a list
```
...a list of `points`?

xymus commented on lib/geometry/polygon.nit:
```diff
@@ -0,0 +1,426 @@
+		points.add_all(pts)
+	end
+
+	# Get an array of x coordinates of the vertices
```
*of the x...

BlackMinou commented:
Re-Reroll : 
* Adapted the bench default values so it takes around 5 seconds for nit benchmarks
* Skip the convex_hull benchmark for Java because it's way too long, more than 10min for the same amount of time ... ( same algo and same implementation as nit, so no idea why this is this long)

xymus commented:
:shipit:

privat commented:
ok to test

privat commented:
+1. On peut avoir un diagrame avec les résultats?
